### PR TITLE
Hubs/Scopes Merge 22 - Combine global, isolation and current scope

### PR DESCRIPTION
**Pull request by adinauer**
#skip-changelog

## :scroll: Description

When creating envelopes/events/transactions/... we now have to use all three scopes (global, isolation and current).

TODO: Breadcrumbs currently do not maintain insertion order across scopes. This will be adressed in a future PR.

## :pencil: Checklist

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry/src/main/java/io/sentry/CombinedScopeView.java [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+
+  @Override
+  public @NotNull PropagationContext withPropagationContext(
+      Scope.@NotNull IWithPropagationContext callback) {
+    return getDefaultWriteScope().withPropagationContext(callback);
+  }
+
+  @Override
+  public @NotNull IScope clone() {
+    // TODO just return a new CombinedScopeView with forked scope?
```
Comment (adinauer) [inline sentry/src/main/java/io/sentry/CombinedScopeView.java:411]: Not quite sure here.

File: sentry/src/main/java/io/sentry/CombinedScopeView.java [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+  }
+
+  @Override
+  public void removeExtra(@NotNull String key) {
+    getDefaultWriteScope().removeExtra(key);
+  }
+
+  @Override
+  public @NotNull Contexts getContexts() {
+    return new CombinedContextsView(
```
[the remark this answers is not in the document]
Comment (adinauer) [reply sentry/src/main/java/io/sentry/CombinedScopeView.java:264]: We should use `CombinedContextsView` as `ScopeType.COMBINED` for cross platform SDKs should get all values and calls `setDevice` and `setOperatingSystem` so handing back a copy that isn't mutable (or rather loses updates) isn't ideal.

File: sentry/src/main/java/io/sentry/CombinedScopeView.java (hunk 1 above)
Comment (adinauer) [reply sentry/src/main/java/io/sentry/CombinedScopeView.java:411]: We're now returning a new `CombinedScopeView` with forked isolation and current scope. See https://github.com/getsentry/sentry-java/pull/3375

File: sentry/src/main/java/io/sentry/CombinedContextsView.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    return globalContexts.getBrowser();
+  }
+
+  @Override
+  public void setBrowser(@NotNull Browser browser) {
+    getDefaultContexts().setBrowser(browser);
+  }
+
+  @Override
+  public @Nullable Device getDevice() {
```
[the remark this answers is not in the document]
Comment (adinauer) [reply sentry/src/main/java/io/sentry/CombinedContextsView.java:104]: Hmm this should only happen if someone calls `.setUser` on global scope so an actual `User` instance is returned from `CombinedScopeView` and then pass that into `.setUser` that goes to current or isolation scope.

Can we ignore this edge case for now and address it, when it actually happens for a customer?

One solution would be to have something like `CombinedContextsView` for user and other fields.

[the remark this answers is not in the document]
Comment (adinauer) [reply sentry/src/main/java/io/sentry/CombinedContextsView.java:104]: Here's an example that should avoid the problem: https://github.com/getsentry/sentry-java/blob/ab1c3a6ccd5557c1620d70606f659fa317bd0476/sentry-android-core/src/main/java/io/sentry/android/core/InternalSentrySdk.java#L87-L98

**Review by lbloder — APPROVED** [review]
The two comments within `Scopes.getGlobalScope()` could be removed:
https://github.com/getsentry/sentry-java/blob/a4744028342b7b0abf7d93e4a1b387cd09a24404/sentry/src/main/java/io/sentry/Scopes.java#L584-L588

LGTM otherwise

**Pull request merged**